dxr/plugins/python/indexers.py

```diff
@@ -63,58 +63,43 @@ def needles_by_line(self):
 
     def _all_needles(self):
         """Return an iterable of needles in (needle name, value, Extent) format."""
-
-        syntax_tree = ast.parse(self.contents.encode('utf-8'))
-
-        # Create a lookup table for ast nodes by lineno and
-        # col_offset, for those that have them.  This makes it easy to
-        # associate nodes with their corresponding token.
-        node_location = {}
-        for node in ast.walk(syntax_tree):
-            if getattr(node, 'lineno', None) is not None:
-                position = (node.lineno, node.col_offset)
-                node_location.setdefault(position, []).append(node)
-
-        tmpfile = StringIO(self.contents)
-        token_gen = tokenize.generate_tokens(tmpfile.readline)
-
-        relocate_nodes = []
+        # AST nodes for classes and functions point to the position of
+        # their 'def' and 'class' tokens. To get the position of their
+        # names, we look for 'def' and 'class' tokens and store the
+        # position of the token immediately following them.
+        node_start_table = {}
+        previous_start = None
+        token_gen = tokenize.generate_tokens(StringIO(self.contents).readline)
         for tok_type, tok_name, start, end, _ in token_gen:
             if tok_type != token.NAME:
-                relocate_nodes = []
                 continue
 
-            cur_token = {'name': tok_name, 'start': start, 'end': end}
-            if relocate_nodes:
-                cur_token['nodes'] = relocate_nodes
-            else:
-                cur_token['nodes'] = node_location.get(start, [])
-
-            relocate_nodes = []
+            if tok_name in ('def', 'class'):
+                previous_start = start
+            elif previous_start:
```

Comment:
Could this fail if previous_start ends up being 0, like if a "def" is the first thing in a file?